src/libDirectoryService/FinalBlockPreProcessing.cpp

```diff
@@ -184,8 +184,46 @@ bool DirectoryService::ComposeFinalBlock() {
       << m_mediator.m_txBlockChain.GetLastBlock().GetHeader().GetBlockNum() + 1
       << "][" << m_finalBlock->GetHeader().GetNumTxs() << "] FINAL");
 
-  LOG_EPOCH(INFO, m_mediator.m_currentEpochNum,
-            "Final block Composed: " << *m_finalBlock);
+  if (m_mediator.m_node->m_prePrepRunning) {
+    LOG_EPOCH(INFO, m_mediator.m_currentEpochNum,
+              "PrePrep Final block Composed: " << *m_finalBlock);
+  } else {
+    LOG_EPOCH(INFO, m_mediator.m_currentEpochNum,
+              "Final block Composed: " << *m_finalBlock);
+  }
+
+  m_completeFinalBlockReady = true;
+  m_cvCompleteFinalBlockReady.notify_all();
+
+  return true;
+}
+
+bool DirectoryService::WaitUntilCompleteFinalBlockIsReady() {
+  LOG_MARKER();
+  unique_lock<mutex> lock(m_mediator.m_node->m_mutexMicroBlock);
```

Comment:
What's the reason for holding the MB mutex during the wait for FB generation?

Reply:
I think it not holding but making sure to wait until the complete DS MB(with txns) is generated before we proceed.